libcaf_net/caf/policy/scribe.hpp

```diff
@@ -0,0 +1,160 @@
+#include <caf/net/receive_policy.hpp>
+#include <caf/net/stream_socket.hpp>
+#include <caf/sec.hpp>
+#include <caf/span.hpp>
+#include <caf/variant.hpp>
+
+#ifdef CAF_WINDOWS
+#  include <winsock2.h>
+#else
+#  include <caf/actor.hpp>
+#  include <sys/socket.h>
+#  include <sys/types.h>
+#endif
+
+namespace caf {
+namespace policy {
+
+class scribe {
+public:
+  explicit scribe(net::stream_socket handle);
+
+  net::stream_socket handle_;
+
+  net::stream_socket handle() {
+    return handle_;
+  }
+
+  template <class Parent>
+  error init(Parent& parent) {
+    prepare_next_read();
+    parent.mask_add(net::operation::read);
+    return none;
+  }
+
+  template <class Parent>
+  bool handle_read_event(Parent& parent) {
+    void* buf = read_buf_.data() + collected_;
+    size_t len = read_threshold_ - collected_;
+    CAF_LOG_TRACE(CAF_ARG(handle_.id) << CAF_ARG(len));
+    variant<size_t, sec> ret = read(handle_, buf, len);
+    if (auto num_bytes = get_if<size_t>(&ret)) {
+      CAF_LOG_DEBUG(CAF_ARG(len) << CAF_ARG(handle_.id) << CAF_ARG(*num_bytes));
+      collected_ += *num_bytes;
+      if (collected_ >= read_threshold_) {
+        parent.application().handle_data(*this, read_buf_);
+        prepare_next_read();
+        return false;
+      } else {
+        return true;
+      }
+    } else {
+      // Make sure WSAGetLastError gets called immediately on Windows.
+      auto err = get<sec>(ret);
+      CAF_LOG_DEBUG("receive failed" << CAF_ARG(err));
+      handle_error(parent, err);
+      return false;
+    }
+  }
+
+  template <class Parent>
+  bool handle_write_event(Parent& parent) {
+    // try to write leftover data!
+    write_some(parent);
+    // get new data from parent
+    for (auto msg = parent.next_message(); msg != nullptr;
+         msg = parent.next_message()) {
+      parent.application().write_message(*this, std::move(msg));
+    }
+    // write prepared data
+    return write_some(parent);
+  }
+
+  template <class Parent>
+  bool write_some(Parent& parent) {
+    if (write_buf_.empty())
+      return false;
+    auto len = write_buf_.size() - written_;
+    void* buf = write_buf_.data() + written_;
+    CAF_LOG_TRACE(CAF_ARG(handle_.id) << CAF_ARG(len));
+    auto ret = net::write(handle_, buf, len);
+    if (auto num_bytes = get_if<size_t>(&ret)) {
+      CAF_LOG_DEBUG(CAF_ARG(len) << CAF_ARG(handle_.id) << CAF_ARG(*num_bytes));
+      // update state
+      written_ += *num_bytes;
+      if (written_ >= write_buf_.size()) {
+        written_ = 0;
+        write_buf_.clear();
+        return false;
+      } else {
+        return true;
+      }
+    } else {
+      CAF_LOG_ERROR("send failed");
+      handle_error(parent, get<sec>(ret));
+      return false;
+    }
+  }
+
+  template <class Parent>
+  void resolve(Parent& parent, const std::string& path, actor listener) {
+    parent.application().resolve(parent, path, listener);
+    // TODO should parent be passed as well?
```

Comment:
I don't understand what this comment is about. We are already passing the `parent` to `resolve`, aren't we?